### PR TITLE
Convert js code with jsdoc as ts

### PR DESCRIPTION
**Pull request by sbrunner**
To use it, use e.g.:
`npm run tots -- --dry --print api/src/Map.js` # to test
`npm run tots -- api/src/Map.js`

Actually, the variable, cast are not completely converted....

Not working:
- convert more than one file in one shot
- all the type cast are not working
- `@type {{ ... }}`
- `function(Type1|Type2): ResultType`

### CODE REVIEW

**Comment by sbrunner** [comment]
Should we remove all the .js in the imports?
I think that we can do that with another transformer to facilitate the script renaming...

**Comment by sbrunner** [comment]
I just notes that `@param {function(angular.IAttributes): string}` is not working...

**Comment by sbrunner** [comment]
Should we do something with the enum?

**Comment by gberaudo** [comment]
> Should we do something with the enum?

Let's try without them and see.

**Comment by ger-benjamin** [comment]
Nice work !

> Should we remove all the .js in the imports?
I think that we can do that with another transformer to facilitate the script renaming...

Yes let do that in another script if that's easier. Or here after the addImport/removeImport functions. As you want. 
We just need enough documentation to know that we have to run two command. (or create one command that run both scripts). That will be useful for projects with custom dev based on ngeo.

> Should we do something with the enum?

It would be great if we can transforms them to typescript enums: https://www.typescriptlang.org/docs/handbook/enums.html

**Comment by sbrunner** [comment]
Ready for review :-)

File: buildtools/codeshift-js2ts.js [hunk 1]
```diff
@@ -0,0 +1,991 @@
...
+      .forEach((path) => {
+        visit('', path.value, (indent, node) => {
+          // if (node.comments) {
+          //   console.log(node.comments[0].value);
+          // }
+          // if (node.innerComments) {
+          //   console.log('innerComments');
+          //   console.log(node.innerComments[0].value);
+          // }
+          // console.log(indent + indent + node.type);
```
Comment (llienher) [inline buildtools/codeshift-js2ts.js:587]: Is it usefull ?

File: docs/codeshift.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# JSCodeShift scripts
+
+We provide tow jscodeshift scripts not migrate the code from JavaScript with types in JSDoc to TypeScript.
```
Comment (llienher) [inline docs/codeshift.md:3]:
```suggestion
We provide two jscodeshift scripts to migrate the code from JavaScript (with types in JSDoc) to TypeScript.
```

File: docs/codeshift.md [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+The first one removes the '.js' in the import, then when we convert a script from `.js` to `.ts` we didn't
+need to change all the relative import.
+Then we should run it first on all the repository with: e.-g.
```
Comment (llienher) [inline docs/codeshift.md:7]:
```suggestion
The first script removes the '.js' in the import. That way, when we convert a script from `.js` to `.ts` we don't
need to change all the relative imports.
We should run it first on all the repository with: e.-g.
```

File: docs/codeshift.md [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+git mv <file_path>.js <file_path>.ts
+npm run cs-tots -- <file_path>.ts
+```
+
+Conversion limitations:
+
+- I use global variable then we shouldn't convert more than one file in one shot.
+- Not all the type cast are working.
+- `@type {{ ... }}` are not working, it should be converted into an `@typedef` first.
+- `function(Type1|Type2): ResultType` is not working the parsing with `doctrine` return an incorrect result.
```
Comment (llienher) [inline docs/codeshift.md:30]:
```suggestion
- I use global variable so we shouldn't convert more than one file in at a time.
- Not all the type casts are working.
- `@type {{ ... }}` is not working, it should be converted into an `@typedef` first.
- `function(Type1|Type2): ResultType` is not working, the parsing with `doctrine` return an incorrect result.
```

File: buildtools/codeshift-js2ts.js [hunk 5]
```diff
@@ -0,0 +1,991 @@
...
+        quoteProps: 'preserve',
+        parser: 'babel',
+      };
+      return prettier.format(result, config);
+    } catch (error) {
+      console.log('Fail to format the result');
+      return result;
+    }
+    //  } catch (error) {
+    //    return result;
```
Comment (llienher) [inline buildtools/codeshift-js2ts.js:985]: Is it normal to catch error in the finally{} ?

**Review by llienher — CHANGES_REQUESTED** [review]
I focused my review mainly on the doc and how to use the scripts. The doc need a few changes to be easily readable. :-)

File: buildtools/codeshift-js2ts.js (hunk 1 above)
Comment (sbrunner) [reply buildtools/codeshift-js2ts.js:587]: To investigate yes :-)

File: buildtools/codeshift-js2ts.js (hunk 5 above)
Comment (sbrunner) [reply buildtools/codeshift-js2ts.js:985]: It's not in the finally, but I remove it...

**Review by ger-benjamin — APPROVED** [review]
Quick review, looks good
Thanks a lot

**Review by llienher — APPROVED** [review]

**Pull request merged**